Lib/socket.py

```diff
@@ -728,6 +728,149 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+    # state.
+    # 2) If set, another socket will be free to bind() on the same
+    # address, effectively preventing this one from accepting connections.
+    # Also, it sets the process in a state where it'll no longer respond
+    # to any signals or graceful kills. The option can still be set though
+    # for SOCK_DGRAM multicast sockets.
+    # See: msdn2.microsoft.com/en-us/library/ms740621(VS.85).aspx
+    iswin = os.name in ('nt', 'cygwin')
+    if reuse_addr is None:
+        reuse_addr = not iswin and hasattr(_socket, 'SO_REUSEADDR')
+    elif reuse_addr and not hasattr(_socket, 'SO_REUSEADDR'):
+        raise ValueError("SO_REUSEADDR not supported on this platform")
+    elif reuse_addr and iswin and type != SOCK_DGRAM:
+        raise ValueError("SO_REUSEADDR allowed for SOCK_DGRAM type only")
+
+    if reuse_port and not hasattr(_socket, "SO_REUSEPORT"):
+        raise ValueError("SO_REUSEPORT not supported on this platform")
+    if type not in (SOCK_STREAM, SOCK_DGRAM):
+        raise ValueError("only SOCK_STREAM and SOCK_DGRAM types are supported")
+    if flags is None:
+        flags = AI_PASSIVE
+    if hybrid_ipv46:
+        if type != SOCK_STREAM:
+            raise ValueError("hybrid_ipv46 requires SOCK_STREAM type")
+        if family not in {AF_INET6, AF_UNSPEC}:
+            raise ValueError("hybrid_ipv46 requires AF_INET6 family")
+        if not supports_hybrid_ipv46():
+            raise ValueError("hybrid_ipv46 not supported on this platform")
+        family = AF_INET6
+
+    info = getaddrinfo(host, port, family, type, 0, flags)
+    if family == AF_UNSPEC:
+        # prefer AF_INET over AF_INET6
+        info.sort(key=lambda x: x[0] == AF_INET, reverse=True)
+
+    err = None
+    for res in info:
+        af, socktype, proto, canonname, sa = res
+        try:
+            sock = socket(af, socktype, proto)
+        except error as _err:
+            err = _err
+            if err.errno == errno.EAFNOSUPPORT:
+                continue
+            else:
+                raise
+        try:
+            if reuse_addr:
+                try:
+                    sock.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)
+                except error:
+                    # Fail later on on bind(), for platforms which may not
+                    # support this option.
+                    pass
+            if reuse_port:
+                sock.setsockopt(SOL_SOCKET, SO_REUSEPORT, 1)
+            if has_ipv6 and af == AF_INET6:
+                if not hybrid_ipv46:
+                    # Disable IPv4/IPv6 dual stack support (enabled by
+                    # default on Linux) which makes a single socket
+                    # listen on both address families in order to be
+                    # consistent across different platforms.
+                    try:
+                        sock.setsockopt(IPPROTO_IPV6, IPV6_V6ONLY, 1)
+                    except NameError:
```

Comment:
Please add a comment explaining that IPPROTO_IPV6 or IPV6_V6ONLY is not available. I was surprised by "except NameError:" here at the first read.